src/main/scala/io/iohk/ethereum/blockchain/sync/FastSyncController.scala

```diff
@@ -0,0 +1,301 @@
+          firstPeer ! PeerActor.Subscribe(Set(BlockHeaders.code))
+          firstPeer ! PeerActor.SendMessage(GetBlockHeaders(Right(targetBlockHash), 1, 0, reverse = false))
+          val timeout = scheduler.scheduleOnce(peerResponseTimeout, self, TargetBlockTimeout)
+          context become waitingForTargetBlock(firstPeer, targetBlockHash, timeout)
+
+        case None =>
+          log.info("Cannot start fast sync, no peers to download from. Scheduling retry in {}", startRetryInterval)
+          scheduleStartFastSync(startRetryInterval, targetBlockHash)
+      }
+  }
+
+  def waitingForTargetBlock(peer: ActorRef,
+                            targetBlockHash: ByteString,
+                            timeout: Cancellable): Receive =handlePeerUpdates orElse {
+    case PeerActor.MessageReceived(blockHeaders: BlockHeaders) =>
+      timeout.cancel()
+      peer ! PeerActor.Unsubscribe
+
+      val targetBlockHeaderOpt = blockHeaders.headers.find(header => header.hash == targetBlockHash)
+      targetBlockHeaderOpt match {
+        case Some(targetBlockHeader) =>
+          log.info("Received target block from peer, starting fast sync")
+
+          scheduler.schedule(0.seconds, printStatusInterval, self, PrintStatus)
+          context become new SyncingHandler(targetBlockHeader).receive
+          self ! EnqueueNodes(Seq(StateMptNodeHash(targetBlockHeader.stateRoot)))
+          self ! ProcessSyncing
+
+        case None =>
+          log.info("Peer ({}) did not respond with target block header, blacklisting and scheduling retry in {}",
+            sender().path.name,
+            startRetryInterval)
+
+          blacklist(sender(), blacklistDuration)
+          scheduleStartFastSync(startRetryInterval, targetBlockHash)
+          context become idle
+      }
+
+    case TargetBlockTimeout =>
+      log.info("Peer ({}) did not respond with target block header (timeout), blacklisting and scheduling retry in {}",
+        sender().path.name,
+        startRetryInterval)
+
+      blacklist(sender(), blacklistDuration)
+      peer ! PeerActor.Unsubscribe
+      scheduleStartFastSync(startRetryInterval, targetBlockHash)
+      context become idle
+  }
+
+  def peersToDownloadFrom: Set[ActorRef] = handshakedPeers.filterNot(isBlacklisted)
+
+  def scheduleStartFastSync(interval: FiniteDuration, targetBlockHash: ByteString): Unit = {
+    scheduler.scheduleOnce(interval, self, StartFastSync(targetBlockHash))
+  }
+
+  def handlePeerUpdates: Receive = {
+    case PeerManagerActor.PeersResponse(peers) =>
+      peers.foreach(_.ref ! PeerActor.GetStatus)
+
+    case PeerActor.StatusResponse(status) =>
+      if (status == PeerActor.Status.Handshaked) {
+        if (!handshakedPeers.contains(sender()) && !isBlacklisted(sender())) {
+          handshakedPeers += sender()
+          context watch sender()
+        }
+      } else removePeer(sender())
+
+    case Terminated(ref) if handshakedPeers.contains(ref) =>
+      removePeer(ref)
```

Comment:
If a child handler is still waiting for response from that peer it will also receive Terminated msg and do `cleanupAndStop` (see FastSyncRequestHandler - it too `watch`es the peer).